core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -152,6 +152,9 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
        */
       updatedBlocks ++=
         blockManager.putIterator(key, values, level, tellMaster = true, effectiveStorageLevel)
+      if (level.useMemory) {
+        blockManager.memoryStore.releasePendingUnrollMemoryForThisThread()
+      }
```

Comment:
@suyanNone this only happens if we failed to unroll in memory. In this case we can't release the pending unroll memory because we created an `Array` that contains some (but not all) of the values in this partitions. We need this `Array` to return an iterator to the caller of this method so we can't release it until after the task ends.